### ISSUE TITLE
Settings for http2 initial stream/connection window sizes

### ISSUE THREAD

**Issue by okhaliavka** [istio/istio#34169]
**Describe the feature request**
Istio should have global setting(s) for `initial_stream_window_size` and `initial_connection_window_size` that go into listeners and http2 clusters that pilot generates. Envoy's default of 256MB is often too much. For us, it's causing trouble with sidecars' memory usage, and decreasing these window sizes helps. We would like to decrease it once and for all connections, but have found that it's impossible, since setting `http2_protocol_options` on a cluster makes envoy think that upstream supports http2. While this is often useful for outbound clusters and works fairly well, this is very problematic for inbound clusters, leaving us no choice but to configure it individually for every http2 workload.

We're not the first ones having problems with http2 flow control defaults:
- https://discuss.istio.io/t/correct-way-to-change-initial-stream-window-size-for-envoy/10061
- https://github.com/istio/istio/issues/15260
- https://github.com/istio/istio/issues/33416

**Describe alternatives you've considered**
`EnvoyFilter`. It works for listeners, but doesn't for clusters for the reasons described above. It could work if EnvoyFilter's `ClusterMatch` allowed to filter clusters with present `http2_protocol_options` though.

[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Comment by ramaraochavali** [comment]
@akhalyavka are you suggesting to have to global default for all or do you need this configurability at individual service?

**Comment by akhalyavka** [comment]
@ramaraochavali Global defaults. Configurability at individual service would be nice too, but it's at least possible with EnvoyFilter.

**Comment by lambdai** [comment]
Still an ongoing issue

**Comment by zirain** [comment]
seems we still need this? cc @hzxuzhonghu

**Comment by ramaraochavali** [comment]
> Configurability at individual service would be nice too, but it's at least possible with EnvoyFilter.

Same is possible for global also with Envoy filter in root namespace without any workload selector.

**Comment by akhalyavka** [comment]
> Same is possible for global also with Envoy filter in root namespace without any workload selector.

@ramaraochavali Setting `http2_protocol_options` on inbound clusters tells envoy to use http2 to talk to the app container. 
Doing it globally breaks all applications that do not support http2.

**Comment by ramaraochavali** [comment]
> Setting http2_protocol_options on inbound clusters tells envoy to use http2 to talk to the app container.
Doing it globally breaks all applications that do not support http2.

Sorry I meant having a workload selector like http2_enabled : true in root namespace and all http2 services have that label. But I agree, having service specific config in DR would be better.

**Comment by keithmattix** [comment]
@jaellio is adding a setting for max_concurrent_streams in istio/api#2952 and I think Zack Butcher is adding some connection pool settings in istio/api#2961. Feels like we're wanting a lot of http2 settings in DestinationRule; do we feel like we need a unified approach to these h2 settings?

**Comment by kyessenov** [comment]
I don't have a strong opinion where to put them. We do need them per-hop (client-server, server-backend) and in both directions.

**Comment by keithmattix** [comment]
+1 on server->app - I've worked with a customer really looking for that functionality

**Comment by ZackButcher** [comment]
IMO this kind of config makes sense in the connection pool settings -- then as Kuat says they'll apply everywhere (via DestinationRule) and with https://github.com/istio/api/pull/2961 they'll be configurable for the server directly too.

That begs the larger question (probably not for this thread), do we want to support other `TrafficPolicy` style configuration via the `Sidecar` as well? [As I called out in this comment](https://github.com/istio/api/pull/2961#issuecomment-1760208500), using the full `TrafficPolicy` in the `Sidecar` has some interesting possibilities for configuring `app-client` and `server-app` (`client-server` is well-handled by DestinationRule today)

**Comment by howardjohn** [comment]
I think we actually implemented TLS in the inbound sidecar api before. it
was removed before it launched; I forget why. We should look into that
before we go down that path though. Shriram wrote it initially

**Comment by kyessenov** [comment]
My feeling is that we probably want to pick a better default. But we need a config setting to opt-out/opt-in during the transition. So we may not need a long term API, but we still need a knob right now.

**Comment by hzxuzhonghu** [comment]
I believe the original design is not to include all the configs from envoy, there are tens(maybe hundreds) of them for only protocol-related

**Comment by kyessenov** [comment]
@hzxuzhonghu Yes that would be the case if the default was good. But it's not - and putting everyone at risk because we don't want to expose a config is not a good strategy.

**Comment by hzxuzhonghu** [comment]
Fair point

**Comment by manderson23** [comment]
Re-opening as it looks like the earlier PRs referenced didn't include the initial stream and connection window sizes
